include/flang/lower/ASTBuilder.h

```diff
@@ -0,0 +1,332 @@
+//===-- lib/lower/ast-builder.h ---------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef FORTRAN_LOWER_AST_BUILDER_H_
+#define FORTRAN_LOWER_AST_BUILDER_H_
+
+#include "flang/parser/parse-tree.h"
+#include "flang/semantics/scope.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace Fortran::lower {
+namespace AST {
+
+struct Evaluation;
+struct Program;
+struct ModuleLikeUnit;
+struct FunctionLikeUnit;
+
+using ParentType =
+    std::variant<Program *, ModuleLikeUnit *, FunctionLikeUnit *, Evaluation *>;
+
+enum class CFGAnnotation {
+  None,
+  Goto,
+  CondGoto,
+  IndGoto,
+  IoSwitch,
+  Switch,
+  Iterative,
+  FirStructuredOp,
+  Return,
+  Terminate
+};
+
+/// Compiler-generated jump
+///
+/// This is used to convert implicit control-flow edges to explicit form in the
+/// decorated AST
+struct CGJump {
+  CGJump(Evaluation *to) : target{to} {}
+  Evaluation *target{nullptr};
+};
+
+/// is `A` a construct (or directive)?
+template <typename A>
+constexpr static bool isConstruct() {
+  return std::is_same_v<A, parser::AssociateConstruct> ||
+         std::is_same_v<A, parser::BlockConstruct> ||
+         std::is_same_v<A, parser::CaseConstruct> ||
+         std::is_same_v<A, parser::ChangeTeamConstruct> ||
+         std::is_same_v<A, parser::CriticalConstruct> ||
+         std::is_same_v<A, parser::DoConstruct> ||
+         std::is_same_v<A, parser::IfConstruct> ||
+         std::is_same_v<A, parser::SelectRankConstruct> ||
+         std::is_same_v<A, parser::SelectTypeConstruct> ||
+         std::is_same_v<A, parser::WhereConstruct> ||
+         std::is_same_v<A, parser::ForallConstruct> ||
+         std::is_same_v<A, parser::CompilerDirective> ||
+         std::is_same_v<A, parser::OpenMPConstruct> ||
+         std::is_same_v<A, parser::OmpEndLoopDirective>;
+}
+
+/// Function-like units can contains lists of evaluations.  These can be
+/// (simple) statements or constructs, where a construct contains its own
+/// evaluations.
+struct Evaluation {
+  using EvalVariant = std::variant<
```

Comment:
Thanks to the templatized aspect of the parse-tree, if a statement is added to the parse-tree and not here, the functions of ASTBuilder.cpp that visit the parse-tree will most likely crash at f18 compile time because it is not possible to create the related `Evaluation` ctor.
So I am not sure it is worth moving this to parse-tree.h, especially since it is very specific to the AST code.
Do you see other use cases for this variant type that would benefit from moving it to parse-tree.h ?

Reply:
Ideally, the classification of action statements,  constructs, construct statements and other statements should be in parse-tree.h and we should be able to say here that an evaluation is action-stmt | constructs | construct-smts | other-stmts | compiler-generated-stmts. But i guess the language currently does not provide for that? 
The only thing specific to here is compiler-generated-stmt(?).

Reply:
I see your point. I will try a bit here to see if I can extract this list automatically from the parse-tree definition like https://github.com/flang-compiler/f18/blob/0c3c39d30e3f166a6a1303337c5fd7eead720fd0/include/flang/parser/parse-tree.h#L466
It will require a bit of templates to get rid of the `common::Indirection`.

Reply:
I don't think everything should be or needs to be in `parse-tree.h`.  That file is already very large and expensive at compile-time (which was a criticism already made elsewhere).

Reply:
If it doesn't relate to the representation of Fortran *syntax*, it shouldn't be in `parse-tree.h`.

Reply:
Thanks @jeanPerier for pointing out ActionStmt in parse-tree.h. It will be great if it can be extracted. But if it is difficult or adds to compile-time overheads then feel free to ignore.
@schweitzpgi , @klausler Is there another place or is your point that this location is fine?

Reply:
@kiranchandramohan I think it's fine here. I say this with the caveat that this code is actively being worked on as I type.

This data structure is, for now, a transitory (and very thin) adapter mapping that only exists to simplify lowering. The future may have other ideas eventually.

Reply:
@kiranchandramohan, I have looked a bit into extracting from the parse-tree  as mentioned above and I think it would be more hassle than having this structure here.

Rational:
Action statements would be ok to extract from the parse-tree `ActionStmt`, but the rest would be much harder. Specifically, the "construct statements" is a subset of the child nodes of `ExecutableConstruct` nodes. This subset does not contain nodes that do not matter to lowering such as `BlockSpecificationPart` because all the related information is inside the scopes at that point. So a manual selection of this subset has to made somewhere. Since, this selection is related to lowering and not to the syntax as @klausler mentioned, then I think it is best for it to live here now.

**For all these reasons, I propose a "won't do" resolution to this conversation. If that is acceptable to you, can you mark this conversation as resolved ?**

Reply:
Thanks @jeanPerier for having a look. Just to complete the discussion, if a person is writing a tool and want to go over all the Action, construct and other statements then that person will have to create all this information by himself or include the flang/lower/ASTBuilder.h class. It would have been better to include a file from the parser directory.

I am fine with your resolution.
I don't see an option to resolve. I think it is limited to the author or people with write access. Please mark as resolved.